File: crypto_screener/coinglass_pairs.py
from __future__ import annotations

import re
from typing import Any


def quote_matches(pair: dict[str, Any], quote_asset: str) -> bool:
    expected = quote_asset.upper()
    return (
        str(pair.get("quote_asset") or "").upper() == expected
        or str(pair.get("settlement_currency") or "").upper() == expected
    )


def pair_symbol_matches_quote(pair: dict[str, Any], quote_asset: str) -> bool:
    expected = quote_asset.upper()
    symbol = str(pair.get("symbol") or "").upper()
    instrument_id = str(pair.get("instrument_id") or "").upper()
    return symbol.endswith(f"/{expected}") or expected in instrument_id


def is_likely_perpetual_instrument(instrument_id: str) -> bool:
    lowered = instrument_id.lower()
    if "perp" in lowered or "swap" in lowered:
        return True
    return re.search(r"[_-]\d{6,8}$", instrument_id) is None


def is_likely_perpetual_pair(pair: dict[str, Any]) -> bool:
    return is_likely_perpetual_instrument(str(pair.get("instrument_id") or ""))


def base_from_pair(pair: dict[str, Any], quote_asset: str = "USDT") -> str:
    symbol = str(pair.get("symbol") or "")
    if "/" in symbol:
        return symbol.split("/", 1)[0].upper()
    instrument_id = str(pair.get("instrument_id") or "").upper()
    return re.sub(r"[^A-Z0-9].*$", "", instrument_id).replace(quote_asset.upper(), "")


def select_price_pair(
    supported_pairs: dict[str, list[dict[str, Any]]],
    exchanges: list[str],
    symbol: str,
    quote_asset: str,
) -> tuple[str, str]:
    expected_symbol = symbol.upper()
    for exchange in exchanges:
        for pair in supported_pairs.get(exchange, []):
            base = str(pair.get("base_asset") or "").upper()
            instrument_id = str(pair.get("instrument_id") or "")
            if base != expected_symbol:
                continue
            if not quote_matches(pair, quote_asset):
                continue
            if not is_likely_perpetual_instrument(instrument_id):
                continue
            return exchange, instrument_id or f"{expected_symbol}{quote_asset.upper()}"
    raise ValueError("no supported configured price pair")

File: apps/api/src/pipeline/coinglassPairs.ts
import type { CoinGlassPair } from '../providers/coinglass.js';

export function quoteMatches(pair: CoinGlassPair, quoteAsset: string): boolean {
  const expected = quoteAsset.toUpperCase();
  return (
    String(pair.quote_asset ?? '').toUpperCase() === expected ||
    String(pair.settlement_currency ?? '').toUpperCase() === expected
  );
}

export function pairSymbolMatchesQuote(pair: CoinGlassPair, quoteAsset: string): boolean {
  const expected = quoteAsset.toUpperCase();
  const symbol = String(pair.symbol ?? '').toUpperCase();
  const instrumentId = String(pair.instrument_id ?? '').toUpperCase();
  return symbol.endsWith(`/${expected}`) || instrumentId.includes(expected);
}

// Non-perpetual only if the id ends in a dated-contract suffix (_YYMMDD/_YYYYMMDD, 6-8 digits) with no "perp"/"swap".
export function isLikelyPerpetualInstrument(instrumentId: string): boolean {
  const lowered = instrumentId.toLowerCase();
  if (lowered.includes('perp') || lowered.includes('swap')) {
    return true;
  }
  return !/[_-]\d{6,8}$/.test(instrumentId);
}

export function isLikelyPerpetualPair(pair: CoinGlassPair): boolean {
  return isLikelyPerpetualInstrument(String(pair.instrument_id ?? ''));
}

export function baseFromPair(pair: CoinGlassPair, quoteAsset = 'USDT'): string {
  const symbol = String(pair.symbol ?? '');
  if (symbol.includes('/')) {
    return (symbol.split('/', 1)[0] as string).toUpperCase();
  }
  const instrumentId = String(pair.instrument_id ?? '').toUpperCase();
  const stripped = instrumentId.replace(/[^A-Z0-9].*$/, '');
  return stripped.split(quoteAsset.toUpperCase()).join('');
}

export function selectPricePair(
  supportedPairs: Record<string, CoinGlassPair[]>,
  exchanges: string[],
  symbol: string,
  quoteAsset: string,
): [exchange: string, contractSymbol: string] {
  const expectedSymbol = symbol.toUpperCase();
  for (const exchange of exchanges) {
    for (const pair of supportedPairs[exchange] ?? []) {
      const base = String(pair.base_asset ?? '').toUpperCase();
      const instrumentId = String(pair.instrument_id ?? '');
      if (base !== expectedSymbol) {
        continue;
      }
      if (!quoteMatches(pair, quoteAsset)) {
        continue;
      }
      if (!isLikelyPerpetualInstrument(instrumentId)) {
        continue;
      }
      return [exchange, instrumentId || `${expectedSymbol}${quoteAsset.toUpperCase()}`];
    }
  }
  throw new Error('no supported configured price pair');
}
